### PR TITLE
Ability to suspend users

### PR DESCRIPTION
**Pull request by or-else**
Only mysql and rethinkdb adapters are updated for new functionality.

### CODE REVIEW

**Comment by or-else** [comment]
@Googlom I've updated both mysql and rethinkdb adapters but I did not touch mongodb adapter. Would you give it a shot or should I try to implement the new functionality in Mongo too?

Here are the changes in RethinkDB: https://github.com/tinode/chat/pull/347/files#diff-113986c41f8310c5ba2d5ee011b9916f

Thanks!

File: server/db/mysql/adapter.go [hunk 1]
```diff
@@ -867,47 +921,68 @@ func (a *adapter) UserDelete(uid t.Uid, hard bool) error {
...
-	rows.Close()
 
-	return uids, err
+	// Change state of p2p topics with the user (p2p topic's owner is 0)
+	if _, err = tx.Exec("UPDATE topics LEFT JOIN subscriptions ON topics.name=subscriptions.topic "+
+		"SET topics.state=?, topics.stateat=? WHERE topics.owner=0 AND subscriptions.userid=?",
+		state, now, decoded_uid); err != nil {
+		return err
+	}
+
```
Comment (aforge) [inline server/db/mysql/adapter.go]: Should we update subscriptions here as well?

File: server/db/mysql/adapter.go [hunk 2]
```diff
@@ -1250,8 +1334,14 @@ func (a *adapter) TopicsForUser(uid t.Uid, keepDeleted bool, opts *t.QueryOpt) (
 				break
 			}
 
+			// Optionally skip deleted topics.
+			if !keepDeleted && top.State == t.StateDeleted {
```
Comment (aforge) [inline server/db/mysql/adapter.go]: Perhaps, modify the query to filter out deleted rows in sql?

File: server/db/mysql/adapter.go [hunk 3]
```diff
@@ -1283,13 +1374,14 @@ func (a *adapter) TopicsForUser(uid t.Uid, keepDeleted bool, opts *t.QueryOpt) (
 			}
 
 			// Optionally skip deleted users.
-			if usr.DeletedAt != nil && !keepDeleted {
+			if !keepDeleted && usr.State == t.StateDeleted {
```
Comment (aforge) [inline server/db/mysql/adapter.go]: Ditto.

File: pbx/model.proto [hunk 4]
```diff
@@ -304,13 +306,15 @@ message TopicDesc {
...
+	string state = 12;
+	int64 state_at = 13;
 }
 
 // MsgTopicSub: topic subscription details, sent in Meta message
 message TopicSub {
 	int64 updated_at = 1;
 	int64 deleted_at = 2;
-
+
```
[the remark this answers is not in the document]
Comment (or-else) [reply pbx/model.proto]: done

File: server/db/mysql/adapter.go [hunk 5]
```diff
@@ -867,47 +921,68 @@ func (a *adapter) UserDelete(uid t.Uid, hard bool) error {
...
-		// TODO: Disable all p2p subscriptions with the user.
-
 		// Disable all subscriptions to topics where the user is the owner.
 		if _, err = tx.Exec("UPDATE subscriptions LEFT JOIN topics ON subscriptions.topic=topics.name "+
-			"SET subscriptions.updatedAt=?, subscriptions.deletedAt=? WHERE topics.owner=?",
+			"SET subscriptions.updatedat=?, subscriptions.deletedat=? WHERE topics.owner=?",
 			now, now, decoded_uid); err != nil {
 			return err
 		}
 		// Disable all topics where the user is the owner.
```
[the remark this answers is not in the document]
Comment (or-else) [reply server/db/mysql/adapter.go]: done

File: server/db/mysql/adapter.go (hunk 1 above)
Comment (or-else) [reply server/db/mysql/adapter.go]: Added a comment that they do not need to be disabled. But that reminded me that I should disable ability to create new subscriptions for suspended users.

File: server/db/mysql/adapter.go (hunk 2 above)
Comment (or-else) [reply server/db/mysql/adapter.go]: done

File: server/db/mysql/adapter.go (hunk 3 above)
Comment (or-else) [reply server/db/mysql/adapter.go]: done

**Comment by or-else** [comment]
I won't merge this right away. I want to disable new subscriptions for disabled users.

**Comment by or-else** [comment]
Mongo adapter also needs to be done before merging.

**Comment by Googlom** [comment]
> @Googlom I've updated both mysql and rethinkdb adapters but I did not touch mongodb adapter. Would you give it a shot or should I try to implement the new functionality in Mongo too?
> 
> Here are the changes in RethinkDB: https://github.com/tinode/chat/pull/347/files#diff-113986c41f8310c5ba2d5ee011b9916f
> 
> Thanks!

Ok. I'll do it soon.

**Comment by or-else** [comment]
I'm merging this to devel but will delay release until mongo is done.

**Pull request merged**